CSharp/Library/Microsoft.Bot.Connector.Shared/MicrosoftAppCredentials.cs

```diff
@@ -161,40 +161,59 @@ public static bool IsTrustedServiceUrl(string serviceUrl)
-            if (!tokenInCache || !TokenNotExpired(oAuthToken) || forceRefresh)
+            Task<OAuthResponse> oAuthTokenTask;
+            OAuthResponse oAuthToken = null;
+            string token = null;
+            bool newAuthTokenTask = false;
+
+            // get tokenTask from cache (it's either null, pending or completed)
+            lock (tokenCache)
             {
-                token = await RefreshAndStoreToken().ConfigureAwait(false);
+                tokenCache.TryGetValue(TokenCacheKey, out oAuthTokenTask);
             }
-            else if (!TokenWithinSafeTimeLimits(oAuthToken))
+
+            // if we have a tokenTask
+            if (oAuthTokenTask != null)
             {
-                string oldToken = token;
-                token = await RefreshAndStoreToken().ConfigureAwait(false);
-                if (string.IsNullOrEmpty(token))
-                {
-                    token = oldToken;
-                }
+                // resolve the task to get the token
+                oAuthToken = await oAuthTokenTask.ConfigureAwait(false);
+                token = oAuthToken.access_token;
             }
-            return token;
-        }
 
-        private async Task<string> RefreshAndStoreToken()
-        {
-            try
+            // lock around evaluating getting a new token so only one updated task will be created at a time
+            lock (tokenCache)
             {
-                OAuthResponse oAuthToken = await RefreshTokenAsync().ConfigureAwait(false);
-                cache.AddOrUpdate(TokenCacheKey, oAuthToken, (key, oldToken) => oAuthToken);
-                return oAuthToken.access_token;
+                // token is missing, expired or force refresh
+                if (token == null || 
+                    TokenNotExpired(oAuthToken) == false || 
+                    forceRefresh || 
+                    TokenWithinSafeTimeLimits(oAuthToken) == false)
+                {
+                    // someone else could have already updated the task under their own lock before we got our lock, get the value again 
+                    tokenCache.TryGetValue(TokenCacheKey, out Task<OAuthResponse> oAuthTokenTask2);
+
+                    // if the cache value hasn't changed, then we are the one to do it.
+                    if (oAuthTokenTask == oAuthTokenTask2)
+                    {
+                        oAuthTokenTask = RefreshTokenAsync();
+                        tokenCache[TokenCacheKey] = oAuthTokenTask;
+
+                        // we can't await in a lock, so we will signal to await the new oAuthTokenTask
+                        newAuthTokenTask = true;
+                    }
+                }
             }
-            catch (OAuthException)
+
+            // if we have new token we need to pull it out again
+            if (newAuthTokenTask)
             {
-                throw;
+                oAuthToken = await oAuthTokenTask.ConfigureAwait(false);

```

Comment:
In the event that oauthTokenTask failed, this await will throw the failure exception for every thread that hits this line.